extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/augmented/interceptor/CaffeineCache.java

```diff
@@ -0,0 +1,74 @@
+package io.quarkus.cache.runtime.augmented.interceptor;
+
+import java.time.Duration;
+import java.util.concurrent.Callable;
+import java.util.function.Function;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+
+public class CaffeineCache {
+
+    private Cache<Object, Object> cache;
+
+    public CaffeineCache(Long maximumSize, Duration expireAfterWrite, Duration expireAfterAccess) {
+        Caffeine<Object, Object> builder = Caffeine.newBuilder();
+        if (maximumSize != null) {
+            builder.maximumSize(maximumSize);
+        }
+        if (expireAfterWrite != null) {
+            builder.expireAfterWrite(expireAfterWrite);
+        }
+        if (expireAfterAccess != null) {
+            builder.expireAfterAccess(expireAfterAccess);
+        }
+        cache = builder.build();
+    }
+
+    public Object get(Object key, Callable<Object> mappingFunction) {
```

Comment:
Can you supply the `mappingFunction` upfront at construction? As in, do you expect/support multiple `@CacheLoad` methods for the same cache?

Spring's AOP adapter uses this same style which all providers must conform to. A common complaint is that it blocks the usage of cache refresh (`refreshAfterWrite(duration)`) which asynchronously reloads an entry, returning the stale value while in-flight, prior to the entry being expired (via `expireAfterXXX`). The cache needs to have the function attached, as a `CacheLoader`, to perform this processing (e.g. on a `Cache.getIfPresent(key)`, flexibility of `CacheLoader.reload(key, oldValue)`). If this will always be the same function (the method invocation) then you offer more flexibility by supplying a `Function<K, V>` upfront, as well as avoid the wrapper per call.

Reply:
Thanks for your review @ben-manes! Question asked in #3306.